core/src/main/java/org/jdbi/v3/core/statement/DefineNamedBindingsStatementCustomizer.java

```diff
@@ -13,20 +13,48 @@
  */
 package org.jdbi.v3.core.statement;
 
+import java.lang.reflect.InvocationHandler;
+import java.lang.reflect.Method;
+import java.lang.reflect.Proxy;
 import java.sql.PreparedStatement;
-import java.sql.SQLException;
 import java.util.Set;
 
-import org.jdbi.v3.core.argument.NullArgument;
+import org.jdbi.v3.core.argument.Argument;
+import org.jdbi.v3.core.internal.exceptions.Unchecked;
 
 class DefineNamedBindingsStatementCustomizer implements StatementCustomizer {
     @Override
-    public void beforeTemplating(PreparedStatement stmt, StatementContext ctx) throws SQLException {
+    public void beforeTemplating(PreparedStatement stmt, StatementContext ctx) {
         final Set<String> alreadyDefined = ctx.getAttributes().keySet();
         final Binding binding = ctx.getBinding();
+        final SetNullHandler handler = new SetNullHandler();
         binding.getNames().stream()
             .filter(name -> !alreadyDefined.contains(name))
             .forEach(name -> binding.findForName(name, ctx).ifPresent(
-                    a -> ctx.define(name, a instanceof NullArgument ? false : true)));
+                    a -> handler.define(name, a, ctx)));
+    }
+
+    private static class SetNullHandler implements InvocationHandler {
+        private final PreparedStatement fakeStmt = (PreparedStatement)
+                Proxy.newProxyInstance(Thread.currentThread().getContextClassLoader(), new Class<?>[] {PreparedStatement.class}, this);
+        private boolean setNull;
+        private boolean setCalled;
+
+        @Override
+        public Object invoke(Object proxy, Method method, Object[] args) {
+            setCalled = method.getName().startsWith("set");
+            boolean argNull = args.length > 1 && args[1] == null;
+            setNull = "setNull".equals(method.getName()) || argNull;
+            return null;
```

Comment:
The only thing I can think that _might_ be a problem is that any argument implementations that depend on e.g. `statement.isWrapperFor()` or `statement.unwrap()` to unwrap a vendor-specific prepared statement might error with `@DefineNamedBindings` where otherwise it might behave fine.

Reply:
Should return a zero / false value for primitive return types, otherwise ClassCastException

Reply:
The `statement.unwrap()` idea isn't going to work, since all the types you would want to unwrap are classes, not interfaces.

Best we can do is provide an informative exception message stating that because `@DefineNamedBindings` uses proxies to determine which fields are non-null, it is incompatible with arguments that rely on `Wrapper.unwrap()`

Reply:
Pushed fix